From fe19aca3d2b526b7f81e974da931a9486d13c2db Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 9 Sep 2026 12:15:42 +1000
Subject: [PATCH] msado15: _Stream_put_Position add validation checks for
 position.

Appears that the stream test use a fixed buffer internally and large
position doesn't enlarge the buffer.

Stopped the resize until a application can be found that requires the functionality.
---
 dlls/msado15/stream.c        |  8 ++++++--
 dlls/msado15/tests/msado15.c | 16 ++++++++++++++++
 2 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/dlls/msado15/stream.c b/dlls/msado15/stream.c
index 50833aecf34..aa6ad80c0d9 100644
--- a/dlls/msado15/stream.c
+++ b/dlls/msado15/stream.c
@@ -194,14 +194,18 @@ static HRESULT resize_buffer( struct stream *stream, LONG size )
 static HRESULT WINAPI stream_put_Position( _Stream *iface, ADO_LONGPTR pos )
 {
     struct stream *stream = impl_from_Stream( iface );
-    HRESULT hr;
 
     TRACE( "%p, %Id\n", stream, pos );
 
     if (stream->state == adStateClosed) return MAKE_ADO_HRESULT( adErrObjectClosed );
     if (pos < 0) return MAKE_ADO_HRESULT( adErrInvalidArgument );
+    if (pos > stream->size)
+    {
+        FIXME("Resizing stream buffer currently not supported\n");
+        if (stream->type == adTypeText) return MAKE_ADO_HRESULT( adErrIllegalOperation );
+        else if (stream->type == adTypeBinary) return MAKE_ADO_HRESULT( adErrInvalidArgument );
+    }
 
-    if ((hr = resize_buffer( stream, stream->pos )) != S_OK) return hr;
     stream->pos = pos;
     return S_OK;
 }
diff --git a/dlls/msado15/tests/msado15.c b/dlls/msado15/tests/msado15.c
index 3178b8aec2f..10d05e71b74 100644
--- a/dlls/msado15/tests/msado15.c
+++ b/dlls/msado15/tests/msado15.c
@@ -2701,6 +2701,11 @@ static void test_Stream(void)
     hr = _Stream_Open( stream, missing, adModeUnknown, adOpenStreamUnspecified, NULL, NULL );
     ok( hr == MAKE_ADO_HRESULT( adErrObjectOpen ), "got %08lx\n", hr );
 
+    mode = 0xdeadbeef;
+    hr = _Stream_get_Mode( stream, &mode );
+    ok( hr == S_OK, "got %08lx\n", hr );
+    ok( mode == adModeReadWrite, "got %u\n", mode );
+
     state = 0xdeadbeef;
     hr = _Stream_get_State( stream, &state );
     ok( hr == S_OK, "got %08lx\n", hr );
@@ -2774,6 +2779,9 @@ static void test_Stream(void)
     ok( hr == S_OK, "got %08lx\n", hr );
     ok( size == 10, "got %Id\n", size );
 
+    hr = _Stream_put_Position( stream, 50 );
+    ok( hr == MAKE_ADO_HRESULT( adErrIllegalOperation ), "got %08lx\n", hr );
+
     hr = _Stream_put_Position( stream, 2 );
     ok( hr == S_OK, "got %08lx\n", hr );
 
@@ -2814,6 +2822,11 @@ static void test_Stream(void)
     hr = _Stream_Open( stream, missing, adModeUnknown, adOpenStreamUnspecified, NULL, NULL );
     ok( hr == S_OK, "got %08lx\n", hr );
 
+    mode = 0xdeadbeef;
+    hr = _Stream_get_Mode( stream, &mode );
+    ok( hr == S_OK, "got %08lx\n", hr );
+    ok( mode == adModeUnknown, "got %u\n", mode );
+
     hr = _Stream_ReadText( stream, adReadAll, &str );
     ok( hr == MAKE_ADO_HRESULT( adErrIllegalOperation ), "got %08lx\n", hr );
 
@@ -2856,6 +2869,9 @@ static void test_Stream(void)
     ok( hr == S_OK, "got %08lx\n", hr );
     ok( pos == 4, "got %Id\n", pos );
 
+    hr = _Stream_put_Position( stream, 50 );
+    ok( hr == MAKE_ADO_HRESULT( adErrInvalidArgument ), "got %08lx\n", hr );
+
     refs = _Stream_Release( stream );
     ok( !refs, "got %ld\n", refs );
 }
-- 
2.53.0

